service/history/timerQueueActiveProcessor.go

```diff
@@ -422,6 +449,79 @@ Update_History_Loop:
 	return ErrMaxAttemptsExceeded
 }
 
+func (t *timerQueueActiveProcessorImpl) processRetryTimer(task *persistence.TimerTaskInfo) error {
+	t.metricsClient.IncCounter(metrics.TimerTaskRetryTimerScope, metrics.TaskRequests)
+	sw := t.metricsClient.StartTimer(metrics.TimerTaskRetryTimerScope, metrics.TaskLatency)
+	defer sw.Stop()
+
+	processFn := func() error {
+		context, release, err0 := t.cache.getOrCreateWorkflowExecution(t.timerQueueProcessorBase.getDomainIDAndWorkflowExecution(task))
+		released := false
+		defer func() {
+			if !released {
+				release()
+			}
+		}()
+		if err0 != nil {
+			return err0
+		}
+		msBuilder, err1 := context.loadWorkflowExecution()
+		if err1 != nil {
+			if _, ok := err1.(*workflow.EntityNotExistsError); ok {
+				// this could happen if this is a duplicate processing of the task, and the execution has already completed.
+				return nil
+			}
+			return err1
+		}
+
+		if !msBuilder.isWorkflowExecutionRunning() {
+			return nil
+		}
+
+		// generate activity task
+		scheduledID := task.EventID
+		ai, running := msBuilder.GetActivityInfo(scheduledID)
+		if !running || task.ScheduleAttempt < int64(ai.Attempt) {
+			return nil
+		}
+
+		var err error
+		domainID := task.DomainID
+		targetDomainID := ai.DomainID
+		execution := workflow.WorkflowExecution{
+			WorkflowId: common.StringPtr(task.WorkflowID),
+			RunId:      common.StringPtr(task.RunID)}
+		taskList := &workflow.TaskList{
+			Name: &ai.TaskList,
+		}
+		scheduleToStartTimeout := ai.ScheduleToStartTimeout
+
+		release() // release earlier as we don't need the lock anymore
+		released = true
+		err = t.historyService.matchingClient.AddActivityTask(nil, &m.AddActivityTaskRequest{
+			DomainUUID:                    common.StringPtr(targetDomainID),
+			SourceDomainUUID:              common.StringPtr(domainID),
+			Execution:                     &execution,
+			TaskList:                      taskList,
+			ScheduleId:                    &scheduledID,
```

Comment:
If you don't include the attempt then how do you dedupe started calls?

Reply:
good catch, will upate

Reply:
as discussed, will leave it as is now. The impact is that if the task delivered multiple times and the previous one already dispatched to an activity worker, and that task already failed, and a new retry is needed, it would schedule a new backoff timer, but the new task delivery will be dispatched with that new attempt.